_posts/2019-04-05-omakase-typescript.markdown

```diff
@@ -0,0 +1,179 @@
+able to see most of the code in one go.
+
+### What is TypeScript?
+
+TypeScript is a programming language - it contains all of JavaScript, and then a bit more. Using our example above,
+let's compare the scripts for "Hello, Danger" in JavaScript vs TypeScript:
+
+```js
+// JavaScript
+var name = "Danger"
+console.log("Hello, " + danger)
+
+// TypeScript
+var name = "Danger"
+console.log("Hello, " + danger)
+
+// Yep, you're not missing something, there's no difference
+```
+
+Due to TypeScript's aim to only _extend_ JavaScript, your normal JavaScript code should work fine with TypeScript.
+The things TypeScript adds to JavaScript are intended to help you be more explicit about what kinds of data are
+used in your code, a bit like Java.
+
+```diff
+- var name = "Danger"
++ var name: string = "Danger"
+console.log("Hello, " + danger)
+```
+
+This extra `: string` allows the reader to be certain that `name` will only be a string. Annotating your variables
+also allows TypeScript to check this for you. This is very useful because keeping track of changes like the type of
+value in a variable seems easy when it's one or two, but once it starts hitting the hundreds, that's a lot to keep
+track of.
+
+Simply speaking, we call these annotations "Types". Hence the name <i>Type</i>Script. The tag-line for TypeScript
+is "JavaScript which scales" which is a statement that these extra type annotations allows you to work on bigger
+projects. This is because you can verify up-front how correct your code is. This means you have less need to
+understand how every change affects the rest of the program.
+
+### Why does Artsy use TypeScript?
+
+Artsy definitely isn't the size of Microsoft! Artsy is about 30 engineers, and Microsoft are about 60,000. However,
+some of our problems are the same. Developers at Artsy build apps which are made up of thousands of files. Any
+single change to one file, could affect any other file. A change to one individual file can affect the behaviour of
+any number of other files, like throwing a pebble into a pond and causing ripples to spread out to the bank.
+
+Typically, the need to ensure there are no bugs is less of a problem for people building websites. Websites are
+easy to make changes to, because if you change the site - everyone gets the update instantly. However, we also
+build our iOS app with JavaScript, and a change to the app requires Apple to review the changes and for someone to
+download the new version from the App Store.
+
+This means that the iOS team needs to have more checks that everything is OK before shipping the app to the world.
+Using TypeScript gives our team the ability to feel good that the changes we have made are only the changes we
+want.
+
+TypeScript isn't the only programming language to tackle the problem of making JavaScript code safer, but it's the
+one with the biggest community, allows people to re-use their JavaScript knowledge and has really good tools to
```

Comment:
More than getting to re-use knowledge, we can re-use it across multiple domains. Web developers can write iOS code, for example. 

There's also how TypeScript makes it easy to migrate codebases gradually over time, which is a huge benefit that might not occur to non-programmers.

Reply:
This one is tricky, I'm currently framing TS vs other transpiled languages which would have  all of the same runtime advantages as TS

Yeah, I've not been able to figure out the migration angle - but I've added a note